cloud/linode/route_controller.go

```diff
@@ -0,0 +1,243 @@
+package linode
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/linode/linodego"
+	"golang.org/x/exp/slices"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	cloudprovider "k8s.io/cloud-provider"
+	"k8s.io/klog/v2"
+)
+
+type routeCache struct {
+	sync.RWMutex
+	routes     map[int][]linodego.InstanceConfig
+	mtx        sync.Mutex
+	lastUpdate time.Time
+	ttl        time.Duration
+}
+
+func (rc *routeCache) refreshRoutes(ctx context.Context, client Client) error {
+	rc.Lock()
+	defer rc.Unlock()
+
+	if time.Since(rc.lastUpdate) < rc.ttl {
+		return nil
+	}
+
+	instances, err := client.ListInstances(ctx, nil)
+	if err != nil {
+		return err
+	}
+
+	rc.routes = make(map[int][]linodego.InstanceConfig, len(instances))
+	wg := sync.WaitGroup{}
+	for _, instance := range instances {
+		wg.Add(1)
+		go func(ctx context.Context, id int) {
+			defer wg.Done()
+			configs, err := client.ListInstanceConfigs(ctx, id, &linodego.ListOptions{})
+			if err != nil {
+				klog.Errorf("Failed fetching instance configs for instance id %d. Error: %s", id, err.Error())
```

Comment:
We don't care if this fails?

Reply:
Thanks, fixed it to return in case there is an error. There won't be any instanceconfig in cache for the instance if the call fails for that instance. If call for one instance fails, we don't want to throw away the other instance calls which might have succeeded. It retries every routeReconciliationPeriod and it will try to build the cache again after cache timeout and will keep retrying. We should be good with these failures.

Reply:
I am ok to return error in case any call when creating cache fails. @avestuk @eljohnson92 thoughts on this? If we return error, its just that cache won't be updated and routes won't get added/deleted for any instance until all calls succeed and cache gets created again.

Reply:
I think as this is implemented if we are going to have a single TTL it is treating all of the config as a single object and should error out fully if any fail. I do think a more resilient implementation might be to have separate TTLs per config so we can update only failed/out of date caches in the future so we don't have to always update the entire cache at once

Reply:
Thanks. I've now changed the implementation to return error if any call fails.